include/nbl/builtin/hlsl/math/functions.hlsl

```diff
@@ -0,0 +1,609 @@
+
+// Copyright (C) 2018-2022 - DevSH Graphics Programming Sp. z O.O.
+// This file is part of the "Nabla Engine".
+// For conditions of distribution and use, see copyright notice in nabla.h
+
+#ifndef _NBL_BUILTIN_HLSL_MATH_FUNCTIONS_INCLUDED_
+#define _NBL_BUILTIN_HLSL_MATH_FUNCTIONS_INCLUDED_
+
+#include <nbl/builtin/hlsl/math/constants.hlsl>
+#include <nbl/builtin/hlsl/common.hlsl>
+
+
+namespace nbl
+{
+namespace hlsl
+{
+namespace math
+{
+
+namespace impl
+{
+
+static float3 reflect_refract(in bool _refract, in float3 I, in float3 N, in float NdotI, in float NdotTorR, in float rcpOrientedEta)
+{    
+    return N*(NdotI*(_refract ? rcpOrientedEta:1.0)+NdotTorR) - I*(_refract ? rcpOrientedEta:1.0);
+}
+
+static uint bitfieldInsert(in uint base, in uint shifted_masked_value, in uint lo, in uint count)
+{
+    const uint hi = base^lo;
+    return (hi<<count)|shifted_masked_value|lo;
+}
+
+}
+
+
+int dot(in int2 a, in int2 b)    { return a.x*b.x+a.y*b.y; }
+uint dot(in uint2 a, in uint2 b) { return a.x*b.x+a.y*b.y; }
+int dot(in int3 a, in int3 b)    { return a.x*b.x+a.y*b.y+a.z*b.z; }
+uint dot(in uint3 a, in uint3 b) { return a.x*b.x+a.y*b.y+a.z*b.z; }
+int dot(in int4 a, in int4 b)    { return a.x*b.x+a.y*b.y+a.z*b.z+a.w*b.w; }
+uint dot(in uint4 a, in uint4 b) { return a.x*b.x+a.y*b.y+a.z*b.z+a.w*b.w; }
+
+
+template<typename vector_t>
+vector_t erf(in vector_t _x)
+{
+    const float a1 = 0.254829592;
+    const float a2 = -0.284496736;
+    const float a3 = 1.421413741;
+    const float a4 = -1.453152027;
+    const float a5 = 1.061405429;
+    const float p  = 0.3275911;
+
+    vector_t _sign = sign(_x);
+    vector_t x = abs(_x);
+    
+    vector_t t = 1.0 / (1.0 + p*x);
+    vector_t y = 1.0 - (((((a5 * t + a4) * t) + a3) * t + a2) * t + a1) * t * exp(-x * x);
+    
+    return _sign*y;
+}
+
+
+
+float erfInv(in float _x)
```

Comment:
note a TODO: in the comment to template on the vector type and use `mix(,,bvec)` to select (conditional execution isntead of branching)